Move SimpleIntervalTestFactory methods to IntervalUtils

magicDGS opened the pull request:
Closes #3771

magicDGS commented:
@jamesemery  - As you reviewed https://github.com/broadinstitute/gatk/pull/3475 and suggested this, do you want to have a look?

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -1376,4 +1376,81 @@ static public int endOfShard(int shardIndex, int shardSize) {
+        }
+        final int contigIndex = dictionary.getSequenceIndex(contig);
+        if (contigIndex == -1) {
+            throw new UserException(String.format("Invalid coordinate: %s:%s-%s (contig not in dictionary)", contig, start, stop));
```
I would personally probably change these to `IllegalArgumentException`, in this class most of the `UserExceptions` seem to be related invalid input files which these methods don't deal with directly. These errors look to mostly be relevant for test files anyway.

magicDGS replied:
It is also if the codepath can be related with an user invalid argument or configuration, but I understand that in this case it can be different. I've done it like this because in the previous class it was expected to throw `UserException`...

magicDGS replied:
Done. Changed everything to `IllegalArgumentException`

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -1376,4 +1376,81 @@ static public int endOfShard(int shardIndex, int shardSize) {
+     * @throws UserException if there was some problem when creating the location.
+     */
+    public static SimpleInterval createOverEntireContig(final SAMSequenceDictionary dictionary, final int contigIndex) {
+        Utils.nonNull(dictionary, "null dictionary");
```
"null dictionary" -> "dictionary is null"

magicDGS replied:
Done.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -1376,4 +1376,81 @@ static public int endOfShard(int shardIndex, int shardSize) {
+     */
+    public static SimpleInterval createOverEntireContig(final SAMSequenceDictionary dictionary, final String contig) {
+        Utils.nonNull(dictionary, "null dictionary");
+        Utils.nonNull(contig, "null contig");
```
"null contig" -> "contig is null"

magicDGS replied:
Done.

jamesemery commented on src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java:
```diff
@@ -1795,4 +1795,100 @@ public void testBasicGenomicSortOfLocatableErrorContigNotInDictionary() throws I
+                {dictionary, "1", 10, -1, UserException.class},
+                {dictionary, "1", 1000, 100, UserException.class},
+                // more than the length
+                {dictionary, "1", 100, dictionary.getSequence("1").getSequenceLength() + 1, UserException.class}
```
Good set of negative tests

magicDGS replied:
Thanks!

jamesemery commented on src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java:
```diff
@@ -1795,4 +1795,100 @@ public void testBasicGenomicSortOfLocatableErrorContigNotInDictionary() throws I
+
+
+    @Test(dataProvider = "createOverEntireContig")
+    public void testCreateOverEntireContig(final SAMSequenceDictionary dictionary, final String contig, final int length) {
```
add `createOverEntireContig` index tests to this method

magicDGS replied:
Already tested through this method, because it delegates on it...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -1376,4 +1376,81 @@ static public int endOfShard(int shardIndex, int shardSize) {
+     */
+    public static SimpleInterval createInterval(final SAMSequenceDictionary dictionary, final String contig, final int start, final int stop) {
+        Utils.nonNull(dictionary, "null dictionary");
+        Utils.nonNull(contig, "null contig");
```
If we change to throw IAE, then it's correct. In the current state it wasn't...

magicDGS replied:
Done.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java:
```diff
@@ -1795,4 +1795,100 @@ public void testBasicGenomicSortOfLocatableErrorContigNotInDictionary() throws I
+
+    @Test(dataProvider = "invalidCreateOverEntireContigArgs")
+    public void testInvalidCreateOverEntireContigArgs(final SAMSequenceDictionary dictionary, final String contig, final Class<? extends Throwable> exceptionClass) {
+        Assert.assertThrows(exceptionClass, () -> IntervalUtils.createOverEntireContig(dictionary, contig));
```
True, that will be cleaner!

magicDGS replied:
Done.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/exome/TargetCollectionUnitTest.java:
```diff
@@ -279,6 +279,7 @@ public TargetCollectionStub(final int numberOfTargets, final Random rdn) {
                 final int targetLength = (int) Math.max(1,slotLength * rdn.nextDouble());
                 final int start = Math.max(nextBasePos,1 + nextBasePos + ((slotLength - targetLength) >> 1));
                 final int stop = start + targetLength;
+                // TODO - this can create intervals with stop > contigLength
```
I haven't look at the logic of this tests, but it looks like it is generating intervals larger than the contig length - @vruano, can you help me to fix this now that the `TargetsToolsTestUtils` is factor out in favor of `IntervalUtils` methods?

magicDGS commented:
@jamesemery - I addressed your comments, but the tests that are failing are due to invalid interval creation within tests that randomly generate targets. I will need the help to the developer of these tests (I think that it is @vruano) to fix the tests without reducing the number of targets generated...

magicDGS commented:
Closing in favor of #4664
